### PR TITLE
[Mwootton] Add ability to log to roctx. Currently just for functions

### PR DESCRIPTION
**Pull request by junliume**
making branch updates easier by pulling branch from wmootton:roctx_logging.

When CI passed, @mwootton : you can either use this one to replace https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2312 or update your PR (I cannot make changes to it)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -682,6 +682,7 @@ enable_clang_tidy(
         -cppcoreguidelines-avoid-capture-default-when-capturing-this
         -cppcoreguidelines-rvalue-reference-param-not-moved
         -readability-avoid-unconditional-preprocessor-if
+        -bugprone-lambda-function-name
```
Comment (CAHEK7) [inline CMakeLists.txt:685]: I guess it's a bit unrelated to initial PR, but probably it's ok.

**Review by CAHEK7 — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (junliume) [reply CMakeLists.txt:685]: @CAHEK7 it is related here: [Clang Tidy Log](http://micimaster.amd.com/blue/rest/organizations/jenkins/pipelines/MLLibs/pipelines/MIOpen/branches/PR-2312/runs/2/nodes/88/log/?start=0)

```
[2023-09-13T06:02:15.863Z] /home/jenkins/workspace/MLLibs_MIOpen_PR-2312/src/solver/gemm_wrw.cpp:212:17: error: inside a lambda, '__FUNCTION__' expands to the name of the function call operator; consider capturing the name of the enclosing function explicitly [bugprone-lambda-function-name,-warnings-as-errors]
[2023-09-13T06:02:15.863Z]                 MIOPEN_LOG_FUNCTION("groupconv, 1x1");
[2023-09-13T06:02:15.863Z]                 ^
[2023-09-13T06:02:15.863Z] /home/jenkins/workspace/MLLibs_MIOpen_PR-2312/src/include/miopen/logger.hpp:326:49: note: expanded from macro 'MIOPEN_LOG_FUNCTION'
[2023-09-13T06:02:15.863Z]             miopen_log_func_ss << "s_api = " << __FUNCTION__ << " | ";                  \
[2023-09-13T06:02:15.863Z]                                                 ^
```

**Comment by junliume** [comment]
Ping @mwootton for review

**Comment by atamazov** [comment]
### Clarification: This adds the ability to annotate the rocTracer records with information similar to what is logged onto console when `MIOPEN_LOG_FUNCTION=1`.

rocTX application code annotation API: https://rocm.docs.amd.com/projects/roctracer/en/latest/roctracer_spec.html#roctx-application-code-annotation-api

File: src/logger.cpp [hunk 2]
```diff
@@ -123,6 +126,11 @@ bool IsLoggingFunctionCalls()
     return miopen::IsEnabled(MIOPEN_ENABLE_LOGGING{}) && !IsLoggingDebugQuiet();
 }
 
+bool IsLoggingToRoctx()
+{
+    return miopen::IsEnabled(MIOPEN_ENABLE_LOGGING_ROCTX{}) && !IsLoggingDebugQuiet();
```
Comment (atamazov) [inline src/logger.cpp:131]: @junliume [optimize] It is likely that `!IsLoggingDebugQuiet()` is not necessary here. Quieting is used during warm-up by the driver. The tracer records related to warm-up will not be annotated, which may confuse the trace readers. And finally, removal of this check would reduce overhead.

Comment (junliume) [reply src/logger.cpp:131]: Ping @mwootton for the updates

**Pull request merged**